gui/petitions.lua

```diff
@@ -0,0 +1,307 @@
+        for _, p in ipairs(a.parties) do
+            for _, e in ipairs(p.entity_ids) do
+                if e == we then
+                    list[#list+1] = getAgreementDetails(a)
+                end
+            end
+        end
+    end
+    
+    return list
+end
+
+local petitions = defclass(petitions, gui.FramedScreen)
+petitions.ATTRS = {
+    frame_style = gui.GREY_LINE_FRAME,
+    frame_title = 'Petitions',
+    frame_width = 20,
+    frame_height = 18,
+    frame_inset = 1,
+    focus_path = 'petitions',
+}
+
+function petitions:init(args)
+    self.start = 1
+    self.list = args.list
+    local lines = {}
+    for _, p in ipairs(args.list) do
+        local width = 0
+        for _, t in ipairs(p.text) do
+            if t == NEWLINE then
+                self.frame_width = math.max(self.frame_width, width + 2)
+                width = 0
+            elseif type(t) == 'string' then
+                width = width + #t + 1
+            else
+                for _, tok in ipairs(t) do
+                    width = width + #tok.text + 1
+                end
+            end
+        end
+        self.frame_width = math.max(self.frame_width, width + 2)
+    end
+    self.frame_height = 18
+    self.frame_width = math.min(df.global.gps.dimx - 2, self.frame_width)
+end
+
+function paint_token(p, t)
+    if type(t) == 'table' then
+        for _, tok in ipairs(t) do
+            p:string(tok.text, tok.pen)
+        end
+    else
+        p:string(t)
+    end
+end
+
+function petitions:onRenderBody(painter)
+    local n = 1
+    local first = false
+    for _, p in ipairs(self.list) do
+        if p.status == 'FULFILLED' and not self.fulfilled then goto continue end
+        for _, t in ipairs(p.text) do
+            if self.start <= n and n <= self.start + self.frame_height then
+                if first then
+                    painter:newline()
+                    n = n + 1
+                    first = false
+                end
+                if t == NEWLINE then
+                    painter:newline()
+                else
+                    paint_token(painter,t)
+                    painter:string(' ')
+                end
+            end
+            if t == NEWLINE then n = n + 1 end
+        end
+        first = true
+    ::continue::
+    end
+    self.max_start = math.max(1, n - self.frame_height + 1)
+    
+    if n > self.frame_height then
+        if self.start > 1 then
+            painter:seek(self.frame_width - 1, 0):string(string.char(24), COLOR_LIGHTCYAN) -- up
+        end
+        if self.start < self.max_start then
+            painter:seek(self.frame_width - 1, self.frame_height - 1):string(string.char(25), COLOR_LIGHTCYAN) -- down
+        end
+    end
+end
+
+function petitions:onRenderFrame(p, frame)
+    petitions.super.onRenderFrame(self, p, frame)
+    p:seek(frame.x1 + 2, frame.y1+frame.height-1):key_string('CUSTOM_F', "toggle fulfilled")
+end
+
+function petitions:onInput(keys)
```

Comment:
If you use a Label widget to display the text, you get scrolling for free

Reply:
I've probably copied it from somewhere...

I've refactored it to use Label now.